spec/expression.dd

```diff
@@ -86,6 +137,85 @@ functions with `extern (D)` $(DDSUBLINK spec/attribute, linkage, linkage).)
     $(BEST_PRACTICE Even though the order of evaluation is well-defined, writing code that
     depends on it is rarely recommended.)
 
+$(H2 $(LNAME2 temporary-lifetime, Lifetime of Temporaries))
+
+$(P Expressions and statements may create and/or consume rvalues. Such values are called
+$(I temporaries) and do not have a name or a visible scope. Their lifetime is managed automatically
+as defined in this section.)
+
+$(P For each evaluation that yields a temporary value, the lifetime of that temporary begins at the
+evaluation point, similarly to creation of a usual named value initialized with an expression.)
+
+$(P Termination of lifetime of temporaries does not obey the customary scoping rules and is defined
+as follows:)
+
+$(UL
+$(LI If:
+$(OL $(LI the full expression has a smallest short-circuit expression $(I expr); and)
+$(LI the temporary is created on the right-hand side of the `&&` or `||` operator; and)
+$(LI the right-hand side is evaluated,))
+then temporary destructors are evaluated right after the right-hand side
+expression has been evaluated and converted to `bool`. Evaluation of destructors proceeds in
+reverse order of construction.)
+
+$(LI For all other cases, the temporaries generated for the purpose of invoking functions are
+deferred to the end of the full expression. The order of destruction is inverse to the order of
+construction.))
+
+$(P If a subexpression of an expression throws an exception, all temporaries created up to the
+evaluation of that subexpression will be destroyed per the rules above. No destructor calls will
+be issued for temporaries not yet constructed.)
+
+$(P Note: An intuition behind these rules is that destructors of temporaries are deferred to the end of full
+expression and in reverse order of construction, with the exception that the right-hand side of
+`&&` and `||` are considered their own full expressions even when part of larger expressions.)
+
+$(P Note: The ternary expression $(I e$(SUBSCRIPT 1) ? e$(SUBSCRIPT 2) : e$(SUBSCRIPT 3)) is not
+a special case although it evaluates expressions conditionally: $(I e$(SUBSCRIPT 1)) and one of
+$(I e$(SUBSCRIPT 2)) and $(I e$(SUBSCRIPT 3)) may create temporaries. Their destructors are inserted
+to the end of the full expression in the reverse order of creation.)
+
+$(P Example:)
+
+$(SPEC_RUNNABLE_EXAMPLE_COMPILE
+---
+import std.stdio;
+struct S
+{
+    int x;
+    this(int n) { x = n; writefln("~S(%s)", x); }
```

Comment:
Seems to be an extra `~` here in front of `S`

Reply:
Fixed